### PR TITLE
Use jndi-dns to obtain default name servers

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

Using reflection to obtain the default name servers may fail in Java9 and also in previous Java versions if a SecurityManager is present.

Modifications:

Try using jndi-dns to obtain default name servers and only try using reflection if this fails.

Result:

Be able to detect default name servers in all cases. Fixes [#6347].

### CODE REVIEW

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java [hunk 1]
```diff
@@ -44,22 +51,48 @@
...
-                if (a != null) {
-                    defaultNameServers.add(new InetSocketAddress(SocketUtils.addressByName(a), DNS_PORT));
+            DirContext ctx = new InitialDirContext(env);
+            String dnsUrls = (String) ctx.getEnvironment().get("java.naming.provider.url");
+            String[] servers = dnsUrls.split(" ");
+            for (String server : servers) {
+                try {
+                    defaultNameServers.add(SocketUtils.socketAddress(new URI(server).getHost(), DNS_PORT));
+                } catch (URISyntaxException e) {
+                    logger.debug("Unable to parse nameserver URI '{}', skipping it.", server, e);
```
Comment (trustin) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java]: How about: `Skipping a malformed nameserver URI: {}` ?

Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java]: done

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java [hunk 2]
```diff
@@ -44,22 +51,48 @@
...
-            @SuppressWarnings("unchecked")
-            final List<String> list = (List<String>) nameservers.invoke(instance);
-            for (String a: list) {
-                if (a != null) {
-                    defaultNameServers.add(new InetSocketAddress(SocketUtils.addressByName(a), DNS_PORT));
+            DirContext ctx = new InitialDirContext(env);
+            String dnsUrls = (String) ctx.getEnvironment().get("java.naming.provider.url");
+            String[] servers = dnsUrls.split(" ");
+            for (String server : servers) {
+                try {
```
Comment (netkins) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java]: MAJOR Extract this nested try block into a separate method. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1141)

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java [hunk 3]
```diff
@@ -44,22 +51,48 @@
 
     static {
         final List<InetSocketAddress> defaultNameServers = new ArrayList<InetSocketAddress>(2);
+
+        // Using jndi-dns to obtain the default name servers.
+        //
+        // See:
+        // - http://docs.oracle.com/javase/8/docs/technotes/guides/jndi/jndi-dns.html
+        // - http://mail.openjdk.java.net/pipermail/net-dev/2017-March/010695.html
+        Hashtable<String, String> env = new Hashtable<String, String>();
```
Comment (netkins) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java]: MAJOR Replace the synchronized class "Hashtable" by an unsynchronized one such as "HashMap". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1149)

Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java]: Not possible as the constructor expects a `Hashtable` 👎

File: resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java (hunk 2 above)
Comment (normanmaurer) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DnsServerAddresses.java]: nope...

**Comment by yschimke** [comment]
Works well on JDK 9 on Mac for me without extra settings.  

One question: Why do this in a static block if unused and also log "Default DNS servers:" even when clients provide alternatives?

**Comment by yschimke** [comment]
The google server definitions would be useful as DnsServerAddressStreamProvider.google

**Comment by yschimke** [comment]
This change conflicts with #6592, building off that PR would be good.

Also generally I think it should be one line to do any of
- use platform configuration or fallback to google
- use platform configuration or fail
- use google
- use a specific configured host list

**Comment by normanmaurer** [comment]
@yschimke I just rebased and will merge it as it is once the CI pass. We can do another PR as a followup if needed.

File: resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsServerAddressStreamProvider.java [hunk 4]
```diff
@@ -47,22 +54,48 @@
...
-            @SuppressWarnings("unchecked")
-            final List<String> list = (List<String>) nameservers.invoke(instance);
-            for (String a: list) {
-                if (a != null) {
-                    defaultNameServers.add(new InetSocketAddress(SocketUtils.addressByName(a), DNS_PORT));
+            DirContext ctx = new InitialDirContext(env);
+            String dnsUrls = (String) ctx.getEnvironment().get("java.naming.provider.url");
+            String[] servers = dnsUrls.split(" ");
+            for (String server : servers) {
+                try {
```
Comment (netkins) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsServerAddressStreamProvider.java]: MAJOR Extract this nested try block into a separate method. [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1141)

File: resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsServerAddressStreamProvider.java [hunk 5]
```diff
@@ -47,22 +54,48 @@
 
     static {
         final List<InetSocketAddress> defaultNameServers = new ArrayList<InetSocketAddress>(2);
+
+        // Using jndi-dns to obtain the default name servers.
+        //
+        // See:
+        // - http://docs.oracle.com/javase/8/docs/technotes/guides/jndi/jndi-dns.html
+        // - http://mail.openjdk.java.net/pipermail/net-dev/2017-March/010695.html
+        Hashtable<String, String> env = new Hashtable<String, String>();
```
Comment (netkins) [inline resolver-dns/src/main/java/io/netty/resolver/dns/DefaultDnsServerAddressStreamProvider.java]: MAJOR Replace the synchronized class "Hashtable" by an unsynchronized one such as "HashMap". [rule](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1149)

**Comment by normanmaurer** [comment]
Cherry-picked into 4.1 (a0fcb72e5d)

**Pull request closed without merging**